scripts/speech_recognition/oomptimizer.py

```diff
@@ -0,0 +1,420 @@
+    schema = [
+        {"type": NeuralType(("B", "T"), AudioSignal()), "seq_length": "input"},
+        {"type": NeuralType(("B",), LengthsType()), "seq_length": "input"},
+        {
+            "type": NeuralType(("B", "T"), LabelsType()),
+            "seq_length": "output",
+            "vocab_size": model.tokenizer.vocab_size,
+        },
+        {"type": NeuralType(("B",), LengthsType()), "seq_length": "output"},
+    ]
+    if isinstance(model, EncDecMultiTaskModel):
+        schema.extend(
+            [{"type": "dummy"}, {"type": "dummy"}]
+        )  # multi-task has 2 extra tensors not needed for batch size tuning
+
+    print("Setting up the optimizers.")
+    optimizer, _ = model.setup_optimization({"name": optimizer_name, "lr": 1e-7, "weight_decay": 0.0})
+
+    is_2d_bucketing = all(
+        isinstance(item, (list, tuple)) and len(item) == 2 and all(isinstance(v, Number) for v in item)
+        for item in buckets
+    )
+
+    def get_max_seq_lens(buckets):
+        # TODO(pzelasko): support text data inputs.
+        if all(isinstance(item, Number) for item in buckets):
+            # Regular bucketing on input sequence length (1D).
+            return [
+                (
+                    compute_num_samples(dur, sampling_rate=16000),  # num_samples
+                    math.ceil(labels_per_second * dur),  # num_labels; might need to go data-driven for optimal tuning
+                )
+                for dur in buckets
+            ]
+        elif is_2d_bucketing:
+            # 2D bucketing on both input and output sequence lengths.
+            return [
+                (
+                    compute_num_samples(dur, sampling_rate=16000),  # num_samples
+                    toks,  # num labels is given by bucket definition in 2D bucketing.
+                )
+                for dur, toks in buckets
+            ]
+        else:
+            raise RuntimeError("The passed buckets list is invalid, we expected either 1D or 2D buckets.")
+
+    print("Starting profiling.")
+    max_seq_lens = get_max_seq_lens(buckets)
+    gen = ProfilingBatchGenerator(schema=schema, start_batch_size=start_batch_size, rel_gap_thresh=threshold)
+    profile = {}
+
+    for bucket, (seq_len_in, seq_len_out) in zip(buckets, max_seq_lens):
+        print(f"The current sequence lengths are: input={seq_len_in} output={seq_len_out}.")
+        gen.reset()
+        batch_idx = 0
+
+        def step():
+            batch = gen(seq_len_in, seq_len_out)
+            oom = False
+            try:
+                print(f"\tCurrent gap: {gen.current_rel_gap}. Attempting shapes: {[b.shape for b in batch]}", end=" ")
+                optimizer.zero_grad()
+                model.training_step(batch, batch_idx)
+                optimizer.step()
+            except torch.cuda.OutOfMemoryError as e:
+                print(f"- OOM!")
+                torch.cuda.memory.empty_cache()
```

Comment:
One notable thing is that, when you do empty_cache() here, some memory will still be occupied by "batch", since there are living references to it. `del batch` before empty_cache() will probably fix that. I'm not sure if this matters.

Secondarily, if I understand correctly, the intermediate activations used to do backpropagation will also "be alive" when you call empty_cache(). I'm not sure if this matters either, but if you are setting these to None anytime you change buckets in your actual training run, but not doing so here, then effectively you will a time during forward and back prop when intermediate activations from two different shapes are "alive".

Another thought is that I am not sure if torch.cuda.memory.empty_cache() is the correct approach here, but that requires more investigation on my part.

Note that I have not run this code at this time!